src/Language/PureScript/TypeChecker/Entailment.hs

```diff
@@ -381,80 +390,93 @@ entails SolverOptions{..} constraint context hints =
-        (_, wrappedTy, _) <- lookupNewtypeConstructor env tyName
-        pure [Constraint nullSourceAnn C.Coercible [] [wrappedTy, b] Nothing]
-      t
-        | (TypeConstructor _ aTyName, _, axs) <- unapplyTypes a
-        , (TypeConstructor _ bTyName, _, bxs) <- unapplyTypes b
-        , not (null axs) && not (null bxs) && aTyName == bTyName -> do
-            -- If both arguments are applications of the same type constructor
-            -- (e.g. @data D a b = D a@ in the constraint
-            -- @Coercible (D a b) (D a' b')@), infer the roles of the type
-            -- constructor's arguments and generate wanted constraints
-            -- appropriately (e.g. here @a@ is representational and @b@ is
-            -- phantom, yielding @Coercible a a'@).
-            let tyRoles = lookupRoles env aTyName
-                k role ax bx = case role of
-                  Nominal
-                    -- If we had first-class equality constraints, we'd just
-                    -- emit one of the form @(a ~ b)@ here and let the solver
-                    -- recurse. Since we don't we must compare the types at
-                    -- this point and fail if they don't match. This likely
-                    -- means there are cases we should be able to handle that
-                    -- we currently can't, but is at least sound.
-                    | ax == bx ->
-                        Just []
-                    | otherwise ->
-                        Nothing
-                  Representational ->
-                    Just [Constraint nullSourceAnn C.Coercible [] [ax, bx] Nothing]
-                  Phantom ->
-                    Just []
-            fmap concat $ sequence $ zipWith3 k tyRoles axs bxs
-        | (TypeConstructor _ tyName, _, xs) <- unapplyTypes t
-        , not $ null xs
-        , Just (tvs, wrappedTy, _) <- lookupNewtypeConstructor env tyName -> do
-            -- If the first argument is a newtype applied to some other types
-            -- (e.g. @newtype T a = T a@ in @Coercible (T X) b@), look up the
-            -- type of its wrapped field and yield a new wanted constraint in
-            -- terms of that type with the type arguments substituted in (e.g.
-            -- @Coercible (T[X/a]) b = Coercible X b@ in the example).
-            let wrappedTySub = replaceAllTypeVars (zip tvs xs) wrappedTy
-            pure [Constraint nullSourceAnn C.Coercible [] [wrappedTySub, b] Nothing]
-      _ ->
+    coercibleWanteds :: Environment -> SourceType -> SourceType -> m (Maybe [SourceConstraint])
+    coercibleWanteds env a b
+      | (TypeConstructor _ aTyName, _, axs) <- unapplyTypes a
+      , (TypeConstructor _ bTyName, _, bxs) <- unapplyTypes b
+      , Just (aTyKind, _) <- M.lookup aTyName $ types env
+      , Just (bTyKind, _) <- M.lookup bTyName $ types env
+      , (aks, kind) <- unapplyKinds aTyKind
+      , (bks, _) <- unapplyKinds bTyKind
+      , length axs < length aks
+      , length bxs < length bks = do
```

Comment:
Can I run my understanding of what's going on here past you to check it's accurate? So `axs` is the list of type arguments that `a` has, and `aks` is the list of kinds of those arguments (and likewise for `b`). We know that `length axs <= length aks`, because otherwise we would have had a kind unification error by now, so the possibilities are that `length axs` might be less than `length aks` (if `a` isn't fully applied), or otherwise it will be equal to `length aks` (if it is fully applied). And again, same for `b`.

I don't think it should be possible that only one of the two arguments is fully applied, right? Otherwise we would have had a kind unification error on line 400? Does it follow that one of these length checks is redundant?

Reply:
That is, I feel like it should follow from `a` and `b` having the same kind that `length aks - length axs = length bks - length bxs`

Reply:
You‘re absolutely right and the `length bxs < length bks` check is indeed redundant now that kinds are checked in `solveCoercible`. I‘ve removed it and also another redundant `not (null bxs)` check in the next guard (if both type constructors have the same name and the first one has no arguments then it would be ill-kinded for the second one to have arguments).